Mathlib/Analysis/NormedSpace/MultipliableUniformlyOn.lean

```diff
@@ -0,0 +1,151 @@
+  simp only [hasSumUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq] at hr
+  refine ((hr K hK).comp_cexp ?_).congr ?_
+  · simp +contextual [← h1 _]
+    exact hg K hK
+  · filter_upwards with s i hi using by simp [exp_sum, fun y ↦ exp_log (hfn K hK i hi y)]
+
+lemma multipliableUniformlyOn_of_clog {f : ι → α → ℂ} {𝔖 : Set (Set α)}
+    (hf : SummableUniformlyOn (fun i x ↦ log (f i x)) 𝔖) (hfn : ∀ K ∈ 𝔖, ∀ x ∈ K, ∀ i, f i x ≠ 0)
+    (hg : ∀ K ∈ 𝔖, BddAbove <| (fun x ↦ (∑' n, log (f n x)).re) '' K) :
+    MultipliableUniformlyOn f 𝔖 :=
+    ⟨_, hasProdUniformlyOn_of_clog hf hfn hg⟩
+
+namespace Summable
+
+variable {R : Type*} [NormedCommRing R] [NormOneClass R] [CompleteSpace R] [TopologicalSpace α]
+  {f : ι → α → R} {K : Set α} {u : ι → ℝ}
+
+lemma hasProdUniformlyOn_one_add (hK : IsCompact K) (hu : Summable u)
+    (h : ∀ᶠ n in cofinite, ∀ x ∈ K, ‖f n x‖ ≤ u n) (hcts : ∀ n, ContinuousOn (f n) K) :
+    HasProdUniformlyOn (fun n x ↦ 1 + f n x) (fun x ↦ ∏' i, (1 + f i x)) {K} := by
+  simp only [hasProdUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq,
+    tendstoUniformlyOn_iff_tendstoUniformly_comp_coe]
+  by_cases hKe : K = ∅
+  · simp [TendstoUniformly, hKe]
+  · haveI hCK : CompactSpace K := isCompact_iff_compactSpace.mp hK
+    haveI hne : Nonempty K := by rwa [Set.nonempty_coe_sort, Set.nonempty_iff_ne_empty]
+    let f' n : C(K, R) := ⟨_, continuousOn_iff_continuous_restrict.mp (hcts n)⟩
+    have hf'_bd : ∀ᶠ (n : ι) in cofinite, ‖f' n‖ ≤ u n := by
+      simp_rw [ContinuousMap.norm_le_of_nonempty]
+      filter_upwards [h] with n hn using fun x ↦ hn x x.2
+    have hM : Multipliable fun i ↦ 1 + f' i := by
+      apply _root_.multipliable_one_add_of_summable
+      apply hu.of_norm_bounded_eventually
+      simpa only [norm_norm] using hf'_bd
+    convert ContinuousMap.tendsto_iff_tendstoUniformly.mp hM.hasProd
+    · aesop
+    · exact funext fun k ↦ ContinuousMap.tprod_apply hM k
+
+lemma multipliableUniformlyOn_one_add (hK : IsCompact K) (hu : Summable u)
+    (h : ∀ᶠ n in cofinite, ∀ x ∈ K, ‖f n x‖ ≤ u n) (hcts : ∀ n, ContinuousOn (f n) K) :
+    MultipliableUniformlyOn (fun n x ↦ 1 + f n x) {K} :=
+    ⟨_, hasProdUniformlyOn_one_add hK hu h hcts⟩
+
+lemma hasProdUniformlyOn_nat_one_add {f : ℕ → α → R} (hK : IsCompact K) {u : ℕ → ℝ}
+    (hu : Summable u) (h : ∀ᶠ n in atTop, ∀ x ∈ K, ‖f n x‖ ≤ u n)
+    (hcts : ∀ n, ContinuousOn (f n) K) :
+    HasProdUniformlyOn (fun n x ↦ 1 + f n x) (fun x ↦ ∏' i, (1 + f i x)) {K} :=
+  hasProdUniformlyOn_one_add hK hu (Nat.cofinite_eq_atTop ▸ h) hcts
+
+lemma multipliableUniformlyOn_nat_one_add {f : ℕ → α → R} (hK : IsCompact K)
+    {u : ℕ → ℝ} (hu : Summable u) (h : ∀ᶠ n in atTop, ∀ x ∈ K, ‖f n x‖ ≤ u n)
+    (hcts : ∀ n, ContinuousOn (f n) K) :
+    MultipliableUniformlyOn (fun n x ↦ 1 + f n x) {K} :=
+  ⟨_, hasProdUniformlyOn_nat_one_add hK hu h hcts⟩
+
+lemma hasProdLocallyUniformlyOn_one_add [LocallyCompactSpace α] (hK : IsOpen K) (hu : Summable u)
```

Comment:
Do we have a lemma saying that `HasProdUniformlyOn` implies `HasProdLocallyUniformlyOn`? If not, then we probably should; and then it ought to reduce this proof, and its nat version, to one-liners.

We should also wrap these lemmas in a section & promote the `LocallyCompactSpace` assumption to a section-level variable.

Reply:
Oh, sorry, you're right of course – I forgot `K` was not necessarily compact here.